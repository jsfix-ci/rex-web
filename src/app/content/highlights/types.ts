import { Highlight, HighlightColorEnum } from '@openstax/highlighter/dist/api';

export type HighlightData = Highlight;
export interface SummaryHighlights {
  [sectionId: string]: {[pageId: string]: HighlightData[]};
}
export interface SummaryFilters {
  chapters: string[];
  colors: HighlightColorEnum[];
}

export interface State {
  myHighlightsOpen: boolean;
  enabled: boolean;
  focused?: string;
  highlights: null | HighlightData[];
  summary: {
    filters: SummaryFilters,
    loading: boolean;
    chapterCounts: {[key: string]: number};
    highlights: SummaryHighlights;
  };
}
