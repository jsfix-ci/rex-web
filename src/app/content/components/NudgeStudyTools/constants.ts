export const spotlightPadding = 1;
