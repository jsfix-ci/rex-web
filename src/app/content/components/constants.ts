import theme from '../../theme';

export const bookBannerDesktopBigHeight = 13;
export const bookBannerDesktopMiniHeight = 7;
export const bookBannerMobileBigHeight = 10.4;
export const bookBannerMobileMiniHeight = 6;

export const sidebarDesktopWidth = 33.5;
export const sidebarMobileWidth = 28.8;
export const sidebarTransitionTime = 300;

export const searchResultsBarDesktopWidth = 43.5;
export const searchResultsBarMobileWidth = 100;

export const toolbarIconColor = theme.color.primary.gray;
export const toolbarDesktopHeight = 5;
export const toolbarMobileHeight = 4;

export const mobileSearchContainerMargin = 1;
export const toolbarSearchInputDesktopHeight = 3.2;
export const toolbarSearchInputMobileHeight = 3.2;
export const toolbarSearchInputMobileWidth = 18.5;
export const toolbarToggleSearchBarTextMobileWidth = 11.2;
export const toolbarMobileSearchWrapperHeight = toolbarSearchInputMobileHeight
+ (mobileSearchContainerMargin * 2) + 0.1;
export const toolbarMobileElementsHeight = bookBannerMobileMiniHeight
+ toolbarMobileHeight
+ toolbarMobileSearchWrapperHeight;

export const contentTextWidth = 82.5;

export const mainContentBackground = '#fff';

export const maxContentGutter = 6;
export const contentWrapperMaxWidth = contentTextWidth + sidebarDesktopWidth + maxContentGutter * 2;
