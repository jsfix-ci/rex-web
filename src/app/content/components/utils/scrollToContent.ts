import scrollTo from 'scroll-to-element';
import { mobileBreak } from '../../../theme';
import {
  bookBannerDesktopMiniHeight,
  bookBannerMobileMiniHeight,
  toolbarDesktopHeight,
  toolbarMobileHeight
} from '../constants';

const pxPerRem = 10;
// height of navbar in px
const desktopOffset = -pxPerRem * (bookBannerDesktopMiniHeight + toolbarDesktopHeight);
const mobileOffset = -pxPerRem * (bookBannerMobileMiniHeight + toolbarMobileHeight);

export default function scrollToContent(elem: Element | string) {
    // innerWidth works best on Chrome
    // check other browsers at https://ryanve.com/lab/dimensions/
    // default to desktopOffset if window is undefined (at least the target is always visible)
    const offset = window && window.innerWidth <= pxPerRem * mobileBreak ? mobileOffset : desktopOffset;
    return scrollTo(elem, {offset});
}
