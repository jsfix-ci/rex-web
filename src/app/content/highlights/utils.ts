import { Book, LinkedArchiveTreeNode, Page } from '../types';
import {
  archiveTreeSectionIsBook,
  archiveTreeSectionIsChapter,
  findArchiveTreeNode,
  flattenArchiveTree,
} from '../utils/archiveTreeUtils';
import { HighlightLocations } from './types';

export const getHighlightLocationsFromBook = (book: Book) => {
  return new Map(flattenArchiveTree(book.tree).filter((section) =>
    (section.parent && archiveTreeSectionIsBook(section.parent))
    || archiveTreeSectionIsChapter(section)).map((s) => [s.id, s]));
};

export const getHighlightLocationForPage = (locations: HighlightLocations, page: Page | LinkedArchiveTreeNode) => {
  let location = locations.get(page.id);

  if (!location) {
    for (const section of locations.values()) {
      if (archiveTreeSectionIsChapter(section) && findArchiveTreeNode(section, page.id)) {
        location = section;
        break;
      }
    }
  }

  return location;
};
