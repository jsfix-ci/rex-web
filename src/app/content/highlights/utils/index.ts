import { ScrollTarget } from '../../../navigation/types';
import { HighlightScrollTarget } from '../types';

export {
  getHighlightLocationFilters,
  getHighlightColorFiltersWithContent,
  getHighlightLocationFilterForPage,
  getHighlightLocationFiltersWithContent,
  sectionIsHighlightLocationFitler
} from './locationFiltersUtils';

export {
  addToTotalCounts,
  updateInTotalCounts,
  removeFromTotalCounts,
  addSummaryHighlight,
  getHighlightByIdFromSummaryHighlights,
  removeSummaryHighlight,
  updateSummaryHighlight,
  updateSummaryHighlightsDependOnFilters,
  getSortedSummaryHighlights
} from './summaryHighlightsUtils';

export const isHighlightScrollTarget = (target: ScrollTarget): target is HighlightScrollTarget => {
  if (target.type === 'highlight' && typeof target.id === 'string') { return true; }
  return false;
};
