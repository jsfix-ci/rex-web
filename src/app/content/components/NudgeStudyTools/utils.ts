import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides } from '../../studyGuides/selectors';
import { toolbarButtonMargin } from '../constants';
import {
  arrowDesktopHeight,
  arrowLeftMargin,
  arrowMobileHeight,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  contentMarginTop,
  cookieNudgeStudyGuidesCounter,
  cookieNudgeStudyGuidesDate,
  cookieNudgeStudyGuidesPageCounter,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  spotlightPadding,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperRight: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const target = useGetStudyToolsTarget();

  React.useEffect(() => {
    if (target) {
      const { top, left, right, height, width } = target.getBoundingClientRect();
      const padding = remsToPx(spotlightPadding);
      const spotlightTopOffset = top - padding;
      const spotlightLeftOffset = left - padding;
      const spotlightHeight = height + (padding * 2);
      const spotlightWidth = width + (padding * 2) - (isMobile ? 0 : remsToPx(toolbarButtonMargin));
      const arrowTopOffset = spotlightTopOffset + spotlightHeight + remsToPx(arrowTopMargin);
      const arrowLeft = spotlightLeftOffset + remsToPx(arrowLeftMargin);
      const contentWrapperTopOffset = arrowTopOffset
        + remsToPx(isMobile ? arrowMobileHeight : arrowDesktopHeight)
        + remsToPx(contentMarginTop);
      const contentWrapperRight = windowWidth - right - padding;
      const closeButtonLeft = left + spotlightWidth;
      const closeButtonTopOffset = contentWrapperTopOffset - remsToPx(closeButtonDistanceFromContent);

      setPositions({
        arrowLeft,
        arrowTopOffset,
        closeButtonLeft,
        closeButtonTopOffset,
        contentWrapperRight,
        contentWrapperTopOffset,
        spotlightHeight,
        spotlightLeftOffset,
        spotlightTopOffset,
        spotlightWidth,
      });
    }
    return () => setPositions(null);
  }, [target, windowWidth, isMobile]);

  return positions;
};

const useGetStudyToolsTarget = () => {
  const document = assertDocument();
  const [target, setTarget] = React.useState<HTMLElement | null>(null);
  const studyGuides = useSelector(hasStudyGuides);

  React.useEffect(() => {
    if (studyGuides) {
      setTarget(document.querySelector(`#${nudgeStudyToolsTargetId}`) as HTMLElement | null);
    }
    return () => setTarget(null);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [studyGuides]);

  return target;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudgeStudyGuidesCounter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudgeStudyGuidesDate);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const now = new Date();
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (now.getTime() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudgeStudyGuidesPageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  Cookies.set(cookieNudgeStudyGuidesPageCounter, (counter + 1).toString());
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudgeStudyGuidesCounter, (counter + 1).toString());
  Cookies.set(cookieNudgeStudyGuidesDate, now.toString());
  Cookies.remove(cookieNudgeStudyGuidesPageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  return React.useEffect(() => {
    const counter = getPageCounterCookie();
    if (page && counter <= nudgeStudyToolsMinPageLimit) { incrementPageCounterCookie(); }
  }, [page]);
};
