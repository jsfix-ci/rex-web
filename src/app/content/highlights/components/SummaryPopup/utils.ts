import { Highlight } from '@openstax/highlighter/dist/api';
import { generateTargetQueryFromScrollTarget } from '../../../../navigation/utils';
import { Book } from '../../../types';
import { findArchiveTreeNode } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const createHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const page = book ? findArchiveTreeNode(book.tree, sourceId) : undefined;
  const target = generateTargetQueryFromScrollTarget({ id, type: 'highlight', elementId });
  return `${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`;
};
