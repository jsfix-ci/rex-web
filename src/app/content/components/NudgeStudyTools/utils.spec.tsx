import { Document, HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestStore from '../../../../test/createTestStore';
import * as reactUtils from '../../../reactUtils';
import { Store } from '../../../types';
import { assertDocument } from '../../../utils';
import * as studyGuidesSelect from '../../studyGuides/selectors';
import { nudgeStudyToolsTargetId } from './constants';
import * as utils from './utils';

describe('useGetStudyToolsTarget', () => {
  let document: Document;
  let target: HTMLElement;
  let store: Store;
  // tslint:disable-next-line: variable-name
  let Component: () => JSX.Element;

  beforeEach(() => {
    document = assertDocument();
    target = document.createElement('div');
    target.setAttribute('id', nudgeStudyToolsTargetId);
    jest.spyOn(target, 'getBoundingClientRect')
      .mockReturnValue({ top: 100, left: 200, right: 300, height: 40, width: 300, bottom: 200 });
    document.body.appendChild(target);

    store = createTestStore();
    Component = () => {
      const targetElement = utils.useGetStudyToolsTarget();
      return <React.Fragment>
        {
          targetElement
          ? `id: ${targetElement.getAttribute('id')}, ${targetElement.toString()}`
          : null
        }
      </React.Fragment>;
    };
  });

  afterEach(() => {
    target.remove();
  });

  it('return target if has study guides', () => {
    jest.spyOn(studyGuidesSelect, 'hasStudyGuides')
      .mockReturnValue(true);

    const component = renderer.create(<Provider store={store}>
      <Component />
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('return null if there are no study guides', () => {
    jest.spyOn(studyGuidesSelect, 'hasStudyGuides')
      .mockReturnValue(false);

    const component = renderer.create(<Provider store={store}>
      <Component />
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('usePositions', () => {
  let document: Document;
  let target: HTMLElement;
  let store: Store;
  // tslint:disable-next-line: variable-name
  let Component: (props: { isMobile: boolean }) => JSX.Element;

  beforeEach(() => {
    document = assertDocument();
    target = document.createElement('div');
    target.setAttribute('id', nudgeStudyToolsTargetId);
    jest.spyOn(target, 'getBoundingClientRect')
      .mockReturnValue({ top: 100, left: 200, right: 300, height: 40, width: 300, bottom: 200 });
    document.body.appendChild(target);

    store = createTestStore();
    Component = (props) => {
      const positions = utils.usePositions(props.isMobile);
      return <React.Fragment>
        {JSON.stringify(positions)}
      </React.Fragment>;
    };
  });

  afterEach(() => {
    target.remove();
  });

  it('return null if target was not found', () => {
    target.remove();

    const component = renderer.create(<Provider store={store}>
      <Component isMobile={false} />
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();

    // assign target so afterEach not throw an error
    target = document.createElement('div');
  });

  it('returns different positions depends on isMobile and windowWidth', () => {
    jest.spyOn(studyGuidesSelect, 'hasStudyGuides')
      .mockReturnValue(true);

    jest.spyOn(reactUtils, 'useDebouncedWindowSize')
      .mockReturnValueOnce([1900])
      .mockReturnValueOnce([1900])
      .mockReturnValueOnce([1900])
      .mockReturnValue([1200]);

    const component = renderer.create(<Provider store={store}>
      <Component isMobile={false} />
    </Provider>);

    component.update(<Provider store={store}>
      <Component isMobile={false} />
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();

    component.update(<Provider store={store}>
      <Component isMobile={true} />
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();

    component.update(<Provider store={store}>
      <Component isMobile={false} />
    </Provider>);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
