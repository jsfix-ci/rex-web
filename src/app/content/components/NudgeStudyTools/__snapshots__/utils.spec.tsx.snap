// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useGetStudyToolsTarget return null if there are no study guides 1`] = `null`;

exports[`useGetStudyToolsTarget return target if has study guides 1`] = `"id: nudge-study-tools, [object HTMLDivElement]"`;

exports[`usePositions return null if target was not found 1`] = `"null"`;

exports[`usePositions returns different positions depends on isMobile and windowWidth 1`] = `"{\\"arrowLeft\\":200,\\"arrowTopOffset\\":160,\\"closeButtonLeft\\":500,\\"closeButtonTopOffset\\":260,\\"contentWrapperRight\\":1590,\\"contentWrapperTopOffset\\":300,\\"spotlightHeight\\":60,\\"spotlightLeftOffset\\":190,\\"spotlightTopOffset\\":90,\\"spotlightWidth\\":300}"`;

exports[`usePositions returns different positions depends on isMobile and windowWidth 2`] = `"{\\"arrowLeft\\":200,\\"arrowTopOffset\\":160,\\"closeButtonLeft\\":520,\\"closeButtonTopOffset\\":192,\\"contentWrapperRight\\":1590,\\"contentWrapperTopOffset\\":232,\\"spotlightHeight\\":60,\\"spotlightLeftOffset\\":190,\\"spotlightTopOffset\\":90,\\"spotlightWidth\\":320}"`;

exports[`usePositions returns different positions depends on isMobile and windowWidth 3`] = `"{\\"arrowLeft\\":200,\\"arrowTopOffset\\":160,\\"closeButtonLeft\\":500,\\"closeButtonTopOffset\\":260,\\"contentWrapperRight\\":890,\\"contentWrapperTopOffset\\":300,\\"spotlightHeight\\":60,\\"spotlightLeftOffset\\":190,\\"spotlightTopOffset\\":90,\\"spotlightWidth\\":300}"`;
